File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.cpp
class Solution
{
    vector<int> res;

public:
    vector<int> restoreArray(vector<vector<int>> &adjacentPairs)
    {
        unordered_map<int, vector<int>> graph;
        for (auto pair : adjacentPairs)
        {
            graph[pair[0]].push_back(pair[1]);
            graph[pair[1]].push_back(pair[0]);
        }

        unordered_set<int> visited;
        for (auto pair : graph)
        {
            if (pair.second.size() == 1)
            {
                DFS(graph, visited, pair.first);
                break;
            }
        }

        return res;
    }

    void DFS(unordered_map<int, vector<int>> &graph, unordered_set<int> &visited, int currentNode)
    {
        visited.insert(currentNode);
        res.push_back(currentNode);

        for (auto &neighbour : graph[currentNode])
        {

            if (visited.count(neighbour) != 0)
                continue;
            DFS(graph, visited, neighbour);
        }
    }
};

File: 1743-restore-the-array-from-adjacent-pairs/1743-restore-the-array-from-adjacent-pairs.java
import java.util.*;

class Solution {
    HashSet<Integer> visited = new HashSet<>();
    List<Integer> result = new ArrayList<>();

    public int[] restoreArray(int[][] adjacentPairs) {
        HashMap<Integer, List<Integer>> graph = buildGraph(adjacentPairs);
        // Find the starting point (a node with only one neighbor)
        int start = 0;
        for (int key : graph.keySet()) {
            if (graph.get(key).size() == 1) {
                start = key;
                break;
            }
        }
        dfs(graph, start);

        // Convert List<Integer> to int[]
        int[] arr = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            arr[i] = result.get(i);
        }
        return arr;
    }

    private void dfs(HashMap<Integer, List<Integer>> graph, int node) {
        if (visited.contains(node)) return;
        visited.add(node);
        result.add(node);
        for (int neighbor : graph.getOrDefault(node, Collections.emptyList())) {
            dfs(graph, neighbor);
        }
    }

    HashMap<Integer, List<Integer>> buildGraph(int[][] adjPairs) {
        HashMap<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : adjPairs) {
            int u = edge[0];
            int v = edge[1];
            graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
        }
        return graph;
    }
}
